python/cudf/cudf/core/reshape.py

```diff
@@ -1160,3 +1160,271 @@ def _length_check_params(obj, columns, name):
+        raise NotImplementedError("normalize is not supported yet")
+
+    if values is None and aggfunc is not None:
+        raise ValueError("aggfunc cannot be used without values.")
+
+    if values is not None and aggfunc is None:
+        raise ValueError("values cannot be used without an aggfunc.")
+
+    if not cudf.api.types.is_nested_list_like(index):
+        index = [index]
+    if not cudf.api.types.is_nested_list_like(columns):
+        columns = [columns]
+
+    rownames = (
+        _get_names(index, rownames, prefix="row") if not rownames else rownames
+    )
+    colnames = (
+        _get_names(columns, colnames, prefix="col")
+        if not colnames
+        else colnames
+    )
+
+    if len(index) != len(rownames):
+        raise ValueError("index and rownames must have same length")
+    if len(columns) != len(colnames):
+        raise ValueError("columns and colnames must have same length")
+
+    if len(set(rownames)) != len(rownames):
+        raise ValueError("rownames must be unique")
+    if len(set(colnames)) != len(colnames):
+        raise ValueError("colnames must be unique")
+
+    data = {
+        **dict(zip(rownames, map(as_column, index))),
+        **dict(zip(colnames, map(as_column, columns))),
+    }
+
+    df = cudf.DataFrame._from_data(data)
+
+    if values is None:
+        df["__dummy__"] = 0
+        kwargs = {"aggfunc": "count", "fill_value": 0}
+    else:
+        df["__dummy__"] = values
+        kwargs = {"aggfunc": aggfunc}
+
+    table = pivot_table(
+        data=df,
+        index=rownames,
+        columns=colnames,
+        values="__dummy__",
+        margins=margins,
+        margins_name=margins_name,
+        dropna=dropna,
+        **kwargs,
+    )
+
+    return table
+
+
+def pivot_table(
+    data,
+    values=None,
+    index=None,
+    columns=None,
+    aggfunc="mean",
+    fill_value=None,
+    margins=None,
+    dropna=None,
+    margins_name=None,
+    observed=None,
+    sort=None,
+):
+    """
+    Create a spreadsheet-style pivot table as a DataFrame.
+
+    Parameters
+    ----------
+    data : DataFrame
+    values : column name or list of column names to aggregate, optional
+    index : list of column names
+            Values to group by in the rows.
+    columns : list of column names
+            Values to group by in the columns.
+    aggfunc : str or dict, default "mean"
+            If dict is passed, the key is column to aggregate
+            and value is function name.
+    fill_value : scalar, default None
+        Value to replace missing values with
+        (in the resulting pivot table, after aggregation).
+    margins : Not supported
+    dropna : Not supported
+    margins_name : Not supported
+    observed : Not supported
+    sort : Not supported
+
+    Returns
+    -------
+    DataFrame
+        An Excel style pivot table.
+    """
+    if margins is not None:
+        raise NotImplementedError("margins is not supported yet")
```

Comment:
For all unsupported arguments, we should match the pandas default value (rather than `None`) and enforce that the argument has the default value, raising a `NotImplementedError` if any other value is passed. The reason for this is that a user _passing the default value explicitly_ should expect their code to work with cuDF because we properly support the default behavior.

Concretely, this means the function should default to `margins = False, margins_name='All', observed=False, sort=True` and fail if those values are not provided.

I would accept the argument that `margins_name` should default to `None` because we cannot use the value of the string unless `margins=True`, which is unsupported. Either way (`None` or `"All"`) is fine with me for that case.

Reply:
Done.